scheduler/src/cook/compute_cluster.clj

```diff
@@ -0,0 +1,124 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.compute-cluster
+  (:require [clojure.tools.logging :as log]
+            [datomic.api :as d]))
+
+(defn- write-compute-cluster
+  "Create a missing compute-cluster for one that's not yet in the database."
+  [conn compute-cluster]
+  (log/info "Installing a new compute cluster in datomic for " compute-cluster)
+  @(d/transact
+     conn
+     [(assoc compute-cluster :db/id (d/tempid :db.part/user))]))
+
+(defn- mesos-cluster->compute-cluster-map-for-datomic
+  "Given a mesos cluster dictionary, determine the datomic entity it should correspond to."
+  [{:keys [compute-cluster-name framework-id]}]
+  {:compute-cluster/type :compute-cluster.type/mesos
+   :compute-cluster/cluster-name compute-cluster-name
+   :compute-cluster/mesos-framework-id framework-id})
+
+(defn get-mesos-cluster-entity-id
+  "Given a configuration map for a mesos cluster, return the datomic entity-id corresponding to the cluster,
+  if it exists. Internal helper function."
+  [unfiltered-db {:keys [compute-cluster-name framework-id]}]
+  {:pre [compute-cluster-name
+         framework-id]}
+  (let [query-result
+        (d/q '[:find [?c]
+               :in $ ?cluster-name? ?mesos-id?
+               :where
+               [?c :compute-cluster/type :compute-cluster.type/mesos]
+               [?c :compute-cluster/cluster-name ?cluster-name?]
+               [?c :compute-cluster/mesos-framework-id ?framework-id?]]
+             unfiltered-db compute-cluster-name framework-id)]
+    (first query-result)))
+
+(defn get-mesos-cluster-map
+  "Process one mesos cluster specification, returning the entity id of the corresponding compute-cluster,
+  creating the cluster if it does not exist."
+  [conn {:keys [compute-cluster-name framework-id] :as mesos-cluster}]
+  {:pre [compute-cluster-name
+         framework-id]}
+  (let [cluster-entity (get-mesos-cluster-entity-id (d/db conn) mesos-cluster)]
+    (when-not cluster-entity
+      (write-compute-cluster conn (mesos-cluster->compute-cluster-map-for-datomic mesos-cluster)))
+    {:cluster-type :mesos-cluster
+     :compute-cluster-name compute-cluster-name
```

Comment:
Why is it `:cluster-type` but `:compute-cluster-name`?

Reply:
Its just a flag at this point, but good catch on the inconsistency.